src/hitchhiker/tree/messaging.cljc

```diff
@@ -0,0 +1,252 @@
+          this-node-index (-> path pop peek)
+          parent (-> path pop pop peek)
+          is-first? (zero? this-node-index)
+          ;;We'll need to find the smallest last-key of the left siblings along the path
+          [left-sibs-on-path is-last?]
+          (loop [path path
+                 is-last? true
+                 left-sibs []]
+            (if (= 1 (count path)) ; are we at the root?
+              [left-sibs is-last?]
+              (let [this-node-index (-> path pop peek)
+                    parent (-> path pop pop peek)
+                    is-first? (zero? this-node-index)
+                    local-last? (= (-> parent :children count dec)
+                                   this-node-index)]
+                (if is-first?
+                  (recur (pop (pop path)) (and is-last? local-last?) left-sibs)
+                  (recur (pop (pop path))
+                         (and is-last? local-last?)
+                         (conj left-sibs
+                               (nth (:children parent)
+                                    (dec this-node-index))))))))
+          left-sibs-min-last (when (seq left-sibs-on-path)
+                               (->> left-sibs-on-path
+                                    (map core/last-key)
+                                    (apply max)))
+          left-sib-filter (if left-sibs-min-last
+                            (drop-while #(>= 0 (core/compare (affects-key %)
+                                                             left-sibs-min-last)))
+                            identity)
+          data-node (peek path)
+          my-last (core/last-key data-node)
+          right-side-filter (if is-last?
+                              identity
+                              (take-while #(>= 0 (core/compare (affects-key %) my-last))))
+          correct-ops (into [] (comp left-sib-filter right-side-filter) ops)
+
+          ;;We include op if leq my left, and not if leq left's left
+          ;;TODO we can't apply all ops, we should ensure to only apply ops whose keys are in the defined range, unless we're the last sibling
+          ]
+      ;(println "left-sibs-min-last" left-sibs-min-last)
+      ;(println "is-last?" is-last?)
+      ;(println "expanding data node" data-node "with ops" correct-ops)
+      (reduce (fn [coll op]
+                (apply-op-to-coll op coll))
+              (:children data-node)
+              correct-ops))))
+
+(defn lookup
+  ([tree key]
+   (lookup tree key nil))
+  ([tree key not-found]
+   (go-try
+     (let [path (<? (core/lookup-path tree key))
+           expanded (apply-ops-in-path path)]
+       (get expanded key not-found)))))
+
+(defn insert
+  [tree key value]
+  (enqueue tree [(assoc (->InsertOp key value)
+                        :tag (uuid)
+                        )]))
+
+(defn delete
+  [tree key]
+  (enqueue tree [(assoc (->DeleteOp key)
+                        :tag (uuid)
+                        )]))
+
+#?(:clj
```

Comment:
I think this needs to be defined for cljs too, but I care see how, if `right-successor` is deferred, it would be challenging. The return value of this function could probably be changed to a channel that streams the iterator, since that would also encapsulate the streaming as backpressure, which would help with latency hiding due to prefetching (whereas now, data is demand-requested via laziness). This would be awesome!

The way to go about this is that the whole `lazy-seq` part should actually spin off single goroutine, which would put the `right-successor` onto the channel in a loop/recur, eagerly walking through the right successors and using the channel backpressure to control the rate (since the `<!` would block otherwise).